File: veloce-backend/app/services/brief_pipeline.py
from datetime import datetime, timezone

from beanie import PydanticObjectId

from app.models.ai_analysis import AiAnalysis
from app.models.assignment import Assignment
from app.models.brief import Brief
from app.models.stage_event import StageEvent
from app.schemas.briefs import IntakeCreate
from app.services.openai_analyze import analyze_brief
from app.util.serialize import analysis_to_out, brief_to_out


async def create_brief_from_intake(
    data: IntakeCreate,
    *,
    actor_label: str,
) -> tuple[Brief, AiAnalysis | None, str | None]:
    now = datetime.now(timezone.utc)
    brief = Brief(
        title=data.title,
        description_rich=data.descriptionRich,
        budget_tier=data.budgetTier,
        timeline_urgency=data.timelineUrgency,
        contact_name=data.contactName,
        contact_email=str(data.contactEmail),
        contact_phone=data.contactPhone,
        stage="new",
        submitted_at=now,
    )
    await brief.insert()

    parsed, err = await analyze_brief(
        title=data.title,
        description_rich=data.descriptionRich,
        budget_tier=data.budgetTier,
        timeline_urgency=data.timelineUrgency,
    )

    analysis: AiAnalysis | None = None
    analysis_error: str | None = err
    if parsed:
        analysis = AiAnalysis(
            brief_id=brief.id,  # type: ignore[arg-type]
            features=parsed.features,
            category=parsed.category,
            effort_hours_min=parsed.effort_hours_min,
            effort_hours_max=parsed.effort_hours_max,
            tech_stack=parsed.tech_stack,
            complexity=parsed.complexity,
        )
        await analysis.insert()
        analysis_error = None

    event = StageEvent(
        brief_id=brief.id,  # type: ignore[arg-type]
        from_stage=None,
        to_stage="new",
        at=now,
        actor_user_id=None,
        actor_name=actor_label,
    )
    await event.insert()

    return brief, analysis, analysis_error


def brief_create_response(brief: Brief, analysis: AiAnalysis | None, analysis_error: str | None):
    from app.schemas.intake_response import CreateBriefResponse

    return CreateBriefResponse(
        brief=brief_to_out(brief),
        analysis=analysis_to_out(analysis) if analysis else None,
        analysisError=analysis_error,
    )


async def reviewer_brief_ids(user_id: str) -> list[PydanticObjectId]:
    assigns = await Assignment.find(Assignment.assigned_to_id == user_id).to_list()
    return list({a.brief_id for a in assigns})

File: veloce-frontend/src/lib/server/brief-pipeline.ts
import type { Db, ObjectId } from "mongodb";

import type { AiAnalysisDoc, BriefDoc } from "./brief-docs";
import { COLLECTIONS } from "./collections";
import { analyzeBrief } from "./openai-analyze";
import { analysisToOut, briefToOut } from "./serialize";
import type { IntakeCreateInput } from "./schemas";

export async function reviewerBriefIds(
  db: Db,
  userId: string,
): Promise<ObjectId[]> {
  const assigns = await db
    .collection(COLLECTIONS.assignments)
    .find({ assigned_to_id: userId })
    .toArray();
  const seen = new Set<string>();
  const out: ObjectId[] = [];
  for (const a of assigns) {
    const bid = a.brief_id as ObjectId;
    const hex = bid.toHexString();
    if (!seen.has(hex)) {
      seen.add(hex);
      out.push(bid);
    }
  }
  return out;
}

export async function createBriefFromIntake(
  db: Db,
  data: IntakeCreateInput,
  actorLabel: string,
): Promise<{
  brief: BriefDoc;
  analysis: AiAnalysisDoc | null;
  analysisError: string | null;
}> {
  const now = new Date();

  const briefInsert = {
    title: data.title,
    description_rich: data.descriptionRich,
    budget_tier: data.budgetTier,
    timeline_urgency: data.timelineUrgency,
    contact_name: data.contactName,
    contact_email: data.contactEmail,
    contact_phone: data.contactPhone,
    stage: "new" as const,
    submitted_at: now,
  };

  const br = await db.collection(COLLECTIONS.briefs).insertOne(briefInsert);
  const briefId = br.insertedId;
  const brief = (await db
    .collection<BriefDoc>(COLLECTIONS.briefs)
    .findOne({ _id: briefId }))!;

  const { parsed, error: analyzeErr } = await analyzeBrief({
    title: data.title,
    descriptionRich: data.descriptionRich,
    budgetTier: data.budgetTier,
    timelineUrgency: data.timelineUrgency,
  });

  let analysis: AiAnalysisDoc | null = null;
  let analysisError: string | null = analyzeErr;
  if (parsed) {
    const aiInsert = {
      brief_id: briefId,
      features: parsed.features,
      category: parsed.category,
      effort_hours_min: parsed.effort_hours_min,
      effort_hours_max: parsed.effort_hours_max,
      tech_stack: parsed.tech_stack,
      complexity: parsed.complexity,
    };
    const aiRes = await db
      .collection(COLLECTIONS.aiAnalyses)
      .insertOne(aiInsert);
    analysis = (await db
      .collection<AiAnalysisDoc>(COLLECTIONS.aiAnalyses)
      .findOne({ _id: aiRes.insertedId }))!;
    analysisError = null;
  }

  const evInsert = {
    brief_id: briefId,
    from_stage: null,
    to_stage: "new" as const,
    at: now,
    actor_user_id: null,
    actor_name: actorLabel,
  };
  await db.collection(COLLECTIONS.stageEvents).insertOne(evInsert);

  return { brief, analysis, analysisError };
}

export function briefCreateResponse(
  brief: BriefDoc,
  analysis: AiAnalysisDoc | null,
  analysisError: string | null,
) {
  return {
    brief: briefToOut(brief),
    analysis: analysis ? analysisToOut(analysis) : null,
    analysisError,
  };
}

/** Replay payload for idempotent intake/webhook retries (analysisError not stored; omitted on replay). */
export async function loadBriefCreateResponse(
  db: Db,
  briefId: ObjectId,
): Promise<ReturnType<typeof briefCreateResponse> | null> {
  const brief = await db
    .collection<BriefDoc>(COLLECTIONS.briefs)
    .findOne({ _id: briefId });
  if (!brief) return null;
  const analysis = await db
    .collection<AiAnalysisDoc>(COLLECTIONS.aiAnalyses)
    .findOne({ brief_id: briefId });
  return briefCreateResponse(brief, analysis ?? null, null);
}

export async function ensureCanView(
  db: Db,
  user: { role: string; _id: ObjectId },
  brief: BriefDoc,
): Promise<boolean> {
  if (user.role === "admin") return true;
  const ids = await reviewerBriefIds(db, user._id.toHexString());
  return ids.some((id) => id.equals(brief._id));
}
